pkg/clusterversion/cv.go

```diff
@@ -64,44 +68,67 @@ func (c *clusterVersionClient) GetClusterVersion() (*configv1.ClusterVersion, er
 	clusterVersion, err := c.GetClusterVersion()
 	if err != nil {
 		return false, err
 	}
 
-	// Move the cluster to the same channel first
 	desired := uc.Spec.Desired
-	if clusterVersion.Spec.Channel != desired.Channel {
-		clusterVersion.Spec.Channel = desired.Channel
-		err = c.client.Update(context.TODO(), clusterVersion)
-		if err != nil {
-			return false, err
+
+	// Commence upgrade using version and channel
+	// Move the cluster to the same channel first
+	if !empty(desired.Channel) && !empty(desired.Version) {
+		if clusterVersion.Spec.Channel != desired.Channel {
+			logger.Info(fmt.Sprintf("Setting ClusterVersion to Channel %s Version %s", desired.Channel, desired.Version))
+			clusterVersion.Spec.Channel = desired.Channel
+			err = c.client.Update(context.TODO(), clusterVersion)
+			if err != nil {
+				return false, err
+			}
+
+			// Retrieve the updated version
+			clusterVersion, err = c.GetClusterVersion()
+			if err != nil {
+				return false, err
+			}
+		}
+
+		// The CVO may need time sync the version before launching the upgrade
+		updateAvailable := false
+		for _, update := range clusterVersion.Status.AvailableUpdates {
+			if update.Version == desired.Version && update.Image != "" {
+				updateAvailable = true
+			}
+		}
+		if !updateAvailable {
+			return false, nil
 		}
 
-		// Retrieve the updated version
-		clusterVersion, err = c.GetClusterVersion()
+		clusterVersion.Spec.Overrides = []configv1.ComponentOverride{}
+		clusterVersion.Spec.DesiredUpdate = &configv1.Update{Version: uc.Spec.Desired.Version}
+		err = c.client.Update(context.TODO(), clusterVersion)
 		if err != nil {
 			return false, err
 		}
 	}
 
-	// The CVO may need time sync the version before launching the upgrade
-	updateAvailable := false
-	for _, update := range clusterVersion.Status.AvailableUpdates {
-		if update.Version == desired.Version && update.Image != "" {
-			updateAvailable = true
+	// Commence upgrade using Image
+	// TODO: Remove dependency on version as per OSD-7609
+	if !empty(desired.Image) && !empty(desired.Version) && empty(desired.Channel) {
+		if clusterVersion.Spec.DesiredUpdate == nil || (clusterVersion.Spec.DesiredUpdate.Image != desired.Image) {
+			logger.Info(fmt.Sprintf("Setting ClusterVersion to Image %s", desired.Image))
+			clusterVersion.Spec.DesiredUpdate = &configv1.Update{Image: desired.Image}
+			err = c.client.Update(context.TODO(), clusterVersion)
+			if err != nil {
+				return false, err
+			}
 		}
 	}
-	if !updateAvailable {
-		return false, nil
-	}
 
-	clusterVersion.Spec.Overrides = []configv1.ComponentOverride{}
-	clusterVersion.Spec.DesiredUpdate = &configv1.Update{Version: uc.Spec.Desired.Version}
-	err = c.client.Update(context.TODO(), clusterVersion)
-	if err != nil {
-		return false, err
+	// If neither (version+channel) nor (image) is defined, throw error.
+	if (empty(desired.Version) && empty(desired.Channel)) && empty(desired.Image) {
```

Comment:
if all three fields have value, the version and channel will take priority?

Reply:
For now, not considering priority anywhere. The `validation` package currently will check that ONLY 1 of (image+version) or (channel+version) is mentioned. This PR would check that in the validation package in pkg/validation/validation.go. I tested it. If all three are mentioned, the validation will throw error.